### PR TITLE
ASoC: tlv320aic32x4: Kernel OOPS while entering DAPM standby mode

### PR DESCRIPTION
**Pull request by b-ak**
During the bootup of the kernel, as soon as the DAPM framework kicks in
it pushes the codec into standy mode.

The existing TVL320AIC32x4 codec driver doesn't prepare the clock in
the probe function.
This leads to an OOPS when the DAPM tries to put it into standy by calling
clk_disable_unprepare()

This patch fixes that problem.

_OOPS Log_
```
[    4.426267] tlv320aic32x4-audio soc:sound: tlv320aic32x4-hifi <-> 3f203000.i2s mapping ok
[    4.539142] aic32x4_set_bias_level: +++++++ 
[    4.539149] aic32x4_set_bias_level: SND_SOC_BIAS_STANDBY 
[    4.544868] ------------[ cut here ]------------
[    4.544901] WARNING: CPU: 3 PID: 123 at drivers/clk/clk.c:594 clk_core_disable+0x170/0x180
[    4.544906] Modules linked in: snd_soc_tlv320aic32x4_i2c snd_soc_tlv320aic32x4 brcmfmac brcmutil cfg80211 rfkill snd_soc_tlv320aic32x4_rpi_soundcard snd_soc_bcm2835_i2s snd_soc_core snd_bcm2835 snd_compress snd_pcm_dmaengine snd_pcm snd_timer snd i2c_bcm2835 spi_bcm2835 bcm2835_gpiomem uio_pdrv_genirq fixed uio i2c_dev fuse ipv6
[    4.545012] CPU: 3 PID: 123 Comm: kworker/u8:2 Not tainted 4.9.14-v7+ #8
[    4.545016] Hardware name: BCM2835
[    4.545030] Workqueue: events_unbound async_run_entry_fn
[    4.545061] [<8010f744>] (unwind_backtrace) from [<8010bd1c>] (show_stack+0x20/0x24)
[    4.545080] [<8010bd1c>] (show_stack) from [<8044709c>] (dump_stack+0xc8/0x114)
[    4.545097] [<8044709c>] (dump_stack) from [<8011ccd0>] (__warn+0xe8/0x114)
[    4.545110] [<8011ccd0>] (__warn) from [<8011cdcc>] (warn_slowpath_null+0x30/0x38)
[    4.545125] [<8011cdcc>] (warn_slowpath_null) from [<804984e4>] (clk_core_disable+0x170/0x180)
[    4.545142] [<804984e4>] (clk_core_disable) from [<80498798>] (clk_core_disable_lock+0x28/0x34)
[    4.545156] [<80498798>] (clk_core_disable_lock) from [<804987d0>] (clk_disable+0x2c/0x30)
[    4.545187] [<804987d0>] (clk_disable) from [<7f48c6e0>] (aic32x4_set_bias_level+0x1a8/0x1d8 [snd_soc_tlv320aic32x4])
[    4.545345] [<7f48c6e0>] (aic32x4_set_bias_level [snd_soc_tlv320aic32x4]) from [<7f1e2554>] (snd_soc_codec_set_bias_level+0x24/0x28 [snd_soc_core])
[    4.545560] [<7f1e2554>] (snd_soc_codec_set_bias_level [snd_soc_core]) from [<7f1e8aec>] (snd_soc_dapm_set_bias_level+0xe4/0x1b0 [snd_soc_core])
[    4.545779] [<7f1e8aec>] (snd_soc_dapm_set_bias_level [snd_soc_core]) from [<7f1eb1c8>] (dapm_pre_sequence_async+0x50/0xb8 [snd_soc_core])
[    4.545897] [<7f1eb1c8>] (dapm_pre_sequence_async [snd_soc_core]) from [<8013ee6c>] (async_run_entry_fn+0x50/0x12c)
[    4.545918] [<8013ee6c>] (async_run_entry_fn) from [<801356c8>] (process_one_work+0x144/0x438)
[    4.545936] [<801356c8>] (process_one_work) from [<80135a00>] (worker_thread+0x44/0x530)
[    4.545953] [<80135a00>] (worker_thread) from [<8013b660>] (kthread+0xe8/0x104)
[    4.545970] [<8013b660>] (kthread) from [<80107ec8>] (ret_from_fork+0x14/0x2c)
[    4.545978] ---[ end trace a084c73a92639691 ]---
[    4.545984] ------------[ cut here ]------------
[    4.545995] WARNING: CPU: 3 PID: 123 at drivers/clk/clk.c:476 clk_core_unprepare+0x16c/0x18c
[    4.545999] Modules linked in: snd_soc_tlv320aic32x4_i2c snd_soc_tlv320aic32x4 brcmfmac brcmutil cfg80211 rfkill snd_soc_tlv320aic32x4_rpi_soundcard snd_soc_bcm2835_i2s snd_soc_core snd_bcm2835 snd_compress snd_pcm_dmaengine snd_pcm snd_timer snd i2c_bcm2835 spi_bcm2835 bcm2835_gpiomem uio_pdrv_genirq fixed uio i2c_dev fuse ipv6
[    4.546087] CPU: 3 PID: 123 Comm: kworker/u8:2 Tainted: G        W       4.9.14-v7+ #8
[    4.546091] Hardware name: BCM2835
[    4.546100] Workqueue: events_unbound async_run_entry_fn
[    4.546119] [<8010f744>] (unwind_backtrace) from [<8010bd1c>] (show_stack+0x20/0x24)
[    4.546133] [<8010bd1c>] (show_stack) from [<8044709c>] (dump_stack+0xc8/0x114)
[    4.546148] [<8044709c>] (dump_stack) from [<8011ccd0>] (__warn+0xe8/0x114)
[    4.546160] [<8011ccd0>] (__warn) from [<8011cdcc>] (warn_slowpath_null+0x30/0x38)
[    4.546174] [<8011cdcc>] (warn_slowpath_null) from [<804981e0>] (clk_core_unprepare+0x16c/0x18c)
[    4.546190] [<804981e0>] (clk_core_unprepare) from [<804992e0>] (clk_unprepare+0x34/0x3c)
[    4.546214] [<804992e0>] (clk_unprepare) from [<7f48c6e8>] (aic32x4_set_bias_level+0x1b0/0x1d8 [snd_soc_tlv320aic32x4])
[    4.546334] [<7f48c6e8>] (aic32x4_set_bias_level [snd_soc_tlv320aic32x4]) from [<7f1e2554>] (snd_soc_codec_set_bias_level+0x24/0x28 [snd_soc_core])
[    4.546557] [<7f1e2554>] (snd_soc_codec_set_bias_level [snd_soc_core]) from [<7f1e8aec>] (snd_soc_dapm_set_bias_level+0xe4/0x1b0 [snd_soc_core])
[    4.546775] [<7f1e8aec>] (snd_soc_dapm_set_bias_level [snd_soc_core]) from [<7f1eb1c8>] (dapm_pre_sequence_async+0x50/0xb8 [snd_soc_core])
[    4.546892] [<7f1eb1c8>] (dapm_pre_sequence_async [snd_soc_core]) from [<8013ee6c>] (async_run_entry_fn+0x50/0x12c)
[    4.546909] [<8013ee6c>] (async_run_entry_fn) from [<801356c8>] (process_one_work+0x144/0x438)
[    4.546926] [<801356c8>] (process_one_work) from [<80135a00>] (worker_thread+0x44/0x530)
[    4.546942] [<80135a00>] (worker_thread) from [<8013b660>] (kthread+0xe8/0x104)
[    4.546958] [<8013b660>] (kthread) from [<80107ec8>] (ret_from_fork+0x14/0x2c)
[    4.546971] ---[ end trace a084c73a92639692 ]---
[    4.546978] aic32x4_set_bias_level: ------- 
```

### CODE REVIEW

**Comment by pelwell** [comment]
This codec driver is currently unmodified from the standard upstream kernel, so we prefer that this change is pushed upstream; once accepted there, we can backport it.

I was going to give you some advice on upstreaming but I found your email address in the maintainers list for this codec, so is this patch already on its way upstream?

**Comment by b-ak** [comment]
Makes sense, so I will send the patch to the mainline first..
I wonder how my mail ID is up there, I haven't pushed it to the mainline yet.. In fact I wasn’t very sure whom to send the patch to, so I pushed it here :)

And post that, I have raspberry pi specific machine driver and overlay, I've written for a hardware(OSHW certification applied) I've designed. That goes in here right ?

**Comment by pelwell** [comment]
> I wonder how my mail ID is up there, I haven't pushed it to the mainline yet.. In fact I wasn’t very sure whom to send the patch to, so I pushed it here :)

This is the output of the get_maintainers.pl script:
```
Liam Girdwood <lgirdwood@gmail.com> (supporter:SOUND - SOC LAYER / DYNAMIC AUDIO POWER MANAGEM...)
Mark Brown <broonie@kernel.org> (supporter:SOUND - SOC LAYER / DYNAMIC AUDIO POWER MANAGEM...)
Jaroslav Kysela <perex@perex.cz> (maintainer:SOUND)
Takashi Iwai <tiwai@suse.com> (maintainer:SOUND)
b-ak <anur.bhargav@gmail.com> (commit_signer:1/1=100%,authored:1/1=100%,added_lines:6/6=100%)
alsa-devel@alsa-project.org (moderated list:SOUND - SOC LAYER / DYNAMIC AUDIO POWER MANAGEM...)
linux-kernel@vger.kernel.org (open list)
```
But on the current rpi-4.14.y it says:
```
Liam Girdwood <lgirdwood@gmail.com> (supporter:SOUND - SOC LAYER / DYNAMIC AUDIO POWER MANAGEM...)
Mark Brown <broonie@kernel.org> (supporter:SOUND - SOC LAYER / DYNAMIC AUDIO POWER MANAGEM...)
Jaroslav Kysela <perex@perex.cz> (maintainer:SOUND)
Takashi Iwai <tiwai@suse.com> (maintainer:SOUND)
alsa-devel@alsa-project.org (moderated list:SOUND - SOC LAYER / DYNAMIC AUDIO POWER MANAGEM...)
linux-kernel@vger.kernel.org (open list)
```

So "commit-signer" must refer to the patch you have submitted here - sorry for the confusion, I've not noticed that it does that before.

However, you now have a list of people to submit the patch to, after you've followed the steps laid out here: https://github.com/raspberrypi/linux/blob/rpi-4.14.y/Documentation/process/submitting-patches.rst

Fortunately, your patch currently meets the submission requirements:
```
$ scripts/checkpatch.pl 0001-ASoC-tlv320aic32x4-Kernel-OOPS-while-entering-DAPM-s.patch 
total: 0 errors, 0 warnings, 12 lines checked

0001-ASoC-tlv320aic32x4-Kernel-OOPS-while-entering-DAPM-s.patch has no obvious style problems and is ready for submission.
```
And being so simple, Mark Brown is likely to approve it very quickly.

**Comment by b-ak** [comment]
I'm so glad to hear that. I'll push it there right away.

And can you please tell me how I should go about the ASoC Machine Driver and the overlay.

I've just started writing the documentation for it. I'm calling it [AudioSense](https://gitlab.com/kakar0t/audiosense-pi)

**Comment by pelwell** [comment]
Just create a Pull Request here for them. One commit for the driver, one for the overlay, and one for any defconfig changes would be good.

**Comment by b-ak** [comment]
Will do that. Thank you!!

File: sound/soc/codecs/tlv320aic32x4.c [hunk 1]
```diff
@@ -1162,6 +1162,12 @@ int aic32x4_probe(struct device *dev, struct regmap *regmap)
 	if (IS_ERR(aic32x4->mclk)) {
 		dev_err(dev, "Failed getting the mclk. The current implementation does not support the usage of this codec without mclk\n");
 		return PTR_ERR(aic32x4->mclk);
+	} else {
```
Comment (6by9) [inline sound/soc/codecs/tlv320aic32x4.c]: I suspect you'll get the comment from upstream that the else and extra braces aren't required as you've ```return```ed within the ```if```.
Just add:
```
	ret = clk_prepare_enable(aic32x4->mclk);
	if (ret != 0) {
		dev_err(dev, "Failed to enable MCLK: %d\n", ret);
		return ret;
	}
```

**Pull request closed without merging**